### PR TITLE
[WIP] Allow the Free Energy Protocol to write multiple lambda states from the Free Energy Protocol

### PR DESCRIPTION
**Pull request by xiki-tempula**
Enable the Free Energy Protocol to write multiple lambda states.

This PR allows the BSS.Protocal._FreeEnergyMixin to have pandas.dataframe as lambda values, which allows multiple lambda scheme to be stored in writing the config.

### CODE REVIEW

**Comment by lohedges** [comment]
Just a note that in order to run your own CI against PR's then you'll need to set up a custom workflow file under (something line) `.github/worflows/exscientia.yaml`, then configure this through your own actions interface. The default `main.yaml` is specific to our main repository, so won't work properly if triggered elsewhere.

Cheers.

**Comment by lohedges** [comment]
That prevents it running at your end, but you wouldn't have a merge accepted back into our devel since it would now break our CI.

**Comment by xiki-tempula** [comment]
@lohedges So you have checked that 
```
  publish:
    if: github.repository_owner == 'michellab'
        build:
```
Would not allow it to run on your CI?

**Comment by lohedges** [comment]
You should be able to disable specific workflows manually from the actions tab, as described [here](https://docs.github.com/en/actions/managing-workflow-runs/disabling-and-enabling-a-workflow). (For some reason it always defaults to being on, which doesn't always make sense on forks.) You could disable ours if your like, then set up your own (with a different name, as suggested above) if you wanted to test the build and run unit tests in `test/Sandpit/Exscientia`, for example. (We could then disable this at the other end if required when merged.)

**Comment by xiki-tempula** [comment]
@lohedges Thanks. I will see later on if we need to set up a different CI.

File: python/BioSimSpace/Sandpit/Exscientia/Protocol/_free_energy_mixin.py [hunk 1]
```diff
@@ -115,138 +120,197 @@ def setPerturbationType(self, perturbation_type):
...
-            if type(lam_vals) is not list:
-                raise TypeError("'lam_vals' must be of type 'list'.")
-
-            # Make sure all lambda values are of type 'float'.
-            if not all(isinstance(x, float) for x in lam_vals):
-                raise TypeError("'lam_vals' must contain values of type 'float'.")
+            if not isinstance(lam_vals, pd.DataFrame):
+                # For pandas < 1.4, TypeError won't be raised if the type
+                # cannot be converted to float
+                lam_vals = pd.DataFrame(data={'fep': lam_vals},
```
Comment (msuruzhon) [inline python/BioSimSpace/Sandpit/Exscientia/Protocol/_free_energy_mixin.py:233]: So all lambdas are stored internally as `DataFrame`? I think tha makes sense.

**Pull request merged**